Fails to compile on macOS

Even by following the [instructions for macOS](https://openfhe-development.readthedocs.io/en/latest/sphinx_rsts/intro/installation/macos.html), openfhe fails to compile on macOS (M2 chip). See below for the error that seems due to a deprecated member.

```
[ 20%] Building CXX object src/core/CMakeFiles/core_tests.dir/unittest/UnitTestSerialize.cpp.o
In file included from /Users/tancrede/git/openfhe-development/src/core/unittest/UnitTestSerialize.cpp:49:
In file included from /Users/tancrede/git/openfhe-development/src/core/include/utils/serial.h:67:
In file included from /Users/tancrede/git/openfhe-development/third-party/cereal/include/cereal/archives/json.hpp:62:
/Users/tancrede/git/openfhe-development/third-party/cereal/include/cereal/external/rapidjson/document.h:102:19: error: 'iterator<std::random_access_iterator_tag, rapidjson::GenericMember<rapidjson::UTF8<>, rapidjson::MemoryPoolAllocator<>>>' is deprecated [-Werror,-Wdeprecated-declarations]
    : public std::iterator<std::random_access_iterator_tag
                  ^
/Users/tancrede/git/openfhe-development/third-party/cereal/include/cereal/external/rapidjson/document.h:548:22: note: in instantiation of template class 'rapidjson::GenericMemberIterator<false, rapidjson::UTF8<>, rapidjson::MemoryPoolAllocator<>>' requested here
    typedef typename GenericMemberIterator<false,Encoding,Allocator>::Iterator MemberIterator;  //!< Member iterator for iterating in object.
                     ^
/Users/tancrede/git/openfhe-development/third-party/cereal/include/cereal/archives/json.hpp:440:15: note: in instantiation of template class 'rapidjson::GenericValue<rapidjson::UTF8<>>' requested here
      typedef JSONValue::ConstMemberIterator MemberIterator;
              ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::random_access_iterator_tag, rapidjson::GenericMember<rapidjson::UTF8<>, rapidjson::MemoryPoolAllocator<>>>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
```

--- a/src/core/include/utils/serial.h
+++ b/src/core/include/utils/serial.h
@@ -46,7 +46,11 @@
 #endif
 #define CEREAL_RAPIDJSON_HAS_CXX11_NOEXCEPT 0
 
-#ifdef __GNUC__
+// In order to correctly identify GCC and clang we must either:
+// 1. use "#if defined(__GNUC__) && !defined(__clang__)" (preferred option)
+// 2. or check the condition "#if defined __clang__" first
+// The reason is: clang always defines __GNUC__ and __GNUC_MINOR__ and __GNUC_PATCHLEVEL__ according to the version of gcc that it claims full compatibility with.
+#if defined(__GNUC__) && !defined(__clang__)
     #if __GNUC__ >= 8
         #pragma GCC diagnostic push
         #pragma GCC diagnostic ignored "-Wclass-memaccess"
@@ -66,7 +70,7 @@
 #include "cereal/types/string.hpp"
 #include "cereal/types/vector.hpp"
 
-#ifdef __GNUC__
+#if defined(__GNUC__) && !defined(__clang__)
     #if __GNUC__ >= 8
         #pragma GCC diagnostic pop
     #endif

--- a/src/core/include/utils/serial.h
+++ b/src/core/include/utils/serial.h
@@ -36,11 +36,7 @@
 #ifndef LBCRYPTO_SERIAL_H
 #define LBCRYPTO_SERIAL_H
 
-// TODO (dsuponit): purge the headers below and combine #pragma for GNU and clang
-#include <fstream>
-#include <iostream>
-#include <sstream>
-#include <string>
+#include "utils/sertype.h"
 
 #ifndef CEREAL_RAPIDJSON_HAS_STDSTRING
     #define CEREAL_RAPIDJSON_HAS_STDSTRING 1
@@ -52,16 +48,14 @@
 
 #ifdef __GNUC__
     #if __GNUC__ >= 8
+        #pragma GCC diagnostic push
         #pragma GCC diagnostic ignored "-Wclass-memaccess"
     #endif
-#endif
-
-#ifdef __clang__
+#elif defined __clang__
     #pragma clang diagnostic push
     #pragma clang diagnostic ignored "-Wunused-private-field"
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
 #endif
-
-#include "utils/sertype.h"
 
 #include "cereal/archives/portable_binary.hpp"
 #include "cereal/archives/json.hpp"
@@ -76,11 +70,14 @@
     #if __GNUC__ >= 8
         #pragma GCC diagnostic pop
     #endif
-#endif
-
-#ifdef __clang__
+#elif defined __clang__
     #pragma clang diagnostic pop
 #endif
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
 
 namespace lbcrypto {
 

--- a/src/core/include/utils/serializable.h
+++ b/src/core/include/utils/serializable.h
@@ -39,7 +39,11 @@
 #endif
 #define CEREAL_RAPIDJSON_HAS_CXX11_NOEXCEPT 0
 
-#ifdef __GNUC__
+// In order to correctly identify GCC and clang we must either:
+// 1. use "#if defined(__GNUC__) && !defined(__clang__)" (preferred option)
+// 2. or check the condition "#if defined __clang__" first
+// The reason is: clang always defines __GNUC__ and __GNUC_MINOR__ and __GNUC_PATCHLEVEL__ according to the version of gcc that it claims full compatibility with.
+#if defined(__GNUC__) && !defined(__clang__)
     #if __GNUC__ >= 8
         #pragma GCC diagnostic push
         #pragma GCC diagnostic ignored "-Wclass-memaccess"
@@ -53,7 +57,7 @@
 #include "cereal/cereal.hpp"
 #include "cereal/types/polymorphic.hpp"
 
-#ifdef __GNUC__
+#if defined(__GNUC__) && !defined(__clang__)
     #if __GNUC__ >= 8
         #pragma GCC diagnostic pop
     #endif

--- a/src/core/include/utils/serializable.h
+++ b/src/core/include/utils/serializable.h
@@ -28,18 +28,8 @@
 // OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //==================================================================================
-
-/*
-  Legacy Serialization utilities
- */
-
 #ifndef LBCRYPTO_SERIALIZABLE_H
 #define LBCRYPTO_SERIALIZABLE_H
-
-// TODO (dsuponit): purge the headers below and combine #pragma for GNU and clang
-#include <iostream>
-#include <string>
-#include <vector>
 
 #ifndef CEREAL_RAPIDJSON_HAS_STDSTRING
     #define CEREAL_RAPIDJSON_HAS_STDSTRING 1
@@ -51,13 +41,13 @@
 
 #ifdef __GNUC__
     #if __GNUC__ >= 8
+        #pragma GCC diagnostic push
         #pragma GCC diagnostic ignored "-Wclass-memaccess"
     #endif
-#endif
-
-#ifdef __clang__
+#elif defined __clang__
     #pragma clang diagnostic push
     #pragma clang diagnostic ignored "-Wunused-private-field"
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
 #endif
 
 #include "cereal/cereal.hpp"
@@ -67,15 +57,15 @@
     #if __GNUC__ >= 8
         #pragma GCC diagnostic pop
     #endif
-#endif
-
-#ifdef __clang__
+#elif defined __clang__
     #pragma clang diagnostic pop
 #endif
 
-namespace lbcrypto {
+#include <iostream>
+#include <string>
+#include <vector>
 
-using Serialized = void*;
+namespace lbcrypto {
 
 /**
  * \class Serializable

--- a/src/core/unittest/UnitTestXallocate.cpp
+++ b/src/core/unittest/UnitTestXallocate.cpp
@@ -114,8 +114,12 @@ void Benchmark(const char* name, AllocFunc allocFunc, DeallocFunc deallocFunc);
 void Benchmark(const char* name, AllocFunc allocFunc, DeallocFunc deallocFunc) {
     TimeVar t1, t_total;
 
-    float ElapsedMicroseconds, TotalElapsedMicroseconds = {0};
-
+    float ElapsedMicroseconds = 0;
+#if defined(__clang__)
+    [[maybe_unused]] float TotalElapsedMicroseconds = 0;
+#else
+    float TotalElapsedMicroseconds = 0;
+#endif
     // Allocate MAX_ALLOCATIONS blocks MAX_BLOCK_SIZE / 2 sized blocks
     TIC(t_total);
     TIC(t1);

